persia/data.py

```diff
@@ -1,50 +1,158 @@
+        self._responder = PyPersiaBatchFlowNatsStubResponder(replica_info, self.sender)
+
+    def __iter__(self):
+        while True:
+            yield None
+
+    def __len__(self) -> int:
+        raise NotImplementedError("StreamingChannel do not implement __len__ function")
+
+
+class PersiaChannel(IterableChannelBase):
+    def __init__(
+        self,
+        buffer_size: int,
+        replica_info: PyPersiaReplicaInfo = None,
+        async_iterator: bool = True,
+    ):
+        super(PersiaChannel, self).__init__(buffer_size, replica_info)
+        self.async_iterator = async_iterator
+
+    def fetch_data(self, sender: PyPersiaBatchDataSender):
+        """Callback function to put the data into PyPersiaBatchDataSender
+
+        Arguments:
+            sender (PyPersiaBatchDataSender): PersiaBatchData sender channel to send the generate data
+                to the PersiaBatchData receive channel
+        """
+        raise NotImplementedError("implement this function to fetch data")
 
     def __iter__(self):
+        # TODO: provide process worker to handler the data with high performance calculating
+        # class WorkerType(Enum):
+        #     THREAD = 1
+        #     PROCESS = 2
+
+        if self.async_iterator:
+            handler = Thread(target=self.fetch_data, args=(self.sender,), daemon=True)
+            handler.start()
+
+        # TODO: process the forward engine error to prevent blocking
+        for _val in range(len(self)):
+            yield _val
+
+        if self.async_iterator:
+            handler.join()
+
+
+class Dataloder(object):
+    r"""Dataloder provide the interface to fetch the PythonBatchData from PyForward
+    wrapper.
+
+    Arguments:
+        channel (IterableChannelBase): data channel for Dataloder to retrive replica info and sender channel
+        forward_buffer_size: (int, optional): gpu forward channel buffer size, this args effect the gpu memory cost
+        is_training (bool, optional): wheter current forward status is training or not
+        timeout (int, optional): timeout for PyFoward to fetch data, millisecond unit
+        num_workers (int, optional): spawn thread worker number for  PyForward to lookup embedding and PythonBatchData prefetch
+        shuffle (bool, optional): iterate the data in order, make the dataflow deterministic
+    """
+
+    def __init__(
+        self,
+        channel: IterableChannelBase,
+        forward_buffer_size: int = 10,
+        is_training: bool = True,
+        timeout: int = 1000 * 60 * 10,
+        num_workers: int = 10,
+        shuffle: bool = False,
+    ):
+        # dynamic import the PyForward due to conditional compilation
+        from persia.prelude import PyForward
+
+        self.channel = channel
+        self.timeout = timeout
+        self.num_workers = num_workers
+
+        self.forward_engine = PyForward(
+            forward_buffer_size,
+            not is_training,
+            not shuffle,
```

Comment:
Is shuffle equal with not in-order? As we are not doing shuffle in our code, our dis-order of data is affected by network transmission and num of data-loader instances. E.g. there is only one  data-loader instance, data will not fully shuffled.